command/volume_status_host.go

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package command
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	humanize "github.com/dustin/go-humanize"
+	"github.com/hashicorp/nomad/api"
+)
+
+func (c *VolumeStatusCommand) hostVolumeStatus(client *api.Client, id, nodeID, nodePool string) int {
+	if id == "" {
+		return c.listHostVolumes(client, nodeID, nodePool)
+	}
+
+	if nodeID != "" || nodePool != "" {
+		c.Ui.Error("-node or -pool options can only be used when no ID is provided")
+		return 1
+	}
+
+	// get a host volume that matches the given prefix or a list of all matches
+	// if an exact match is not found. note we can't use the shared getByPrefix
+	// helper here because the List API doesn't match the required signature
+
+	volStub, possible, err := c.getByPrefix(client, id)
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Error listing volumes: %s", err))
+		return 1
+	}
+	if len(possible) > 0 {
+		out, err := c.formatHostVolumes(possible)
+		if err != nil {
+			c.Ui.Error(fmt.Sprintf("Error formatting: %s", err))
+			return 1
+		}
+		c.Ui.Error(fmt.Sprintf("Prefix matched multiple volumes\n\n%s", out))
+		return 1
+	}
+
+	vol, _, err := client.HostVolumes().Get(volStub.ID, nil)
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Error querying volume: %s", err))
+		return 1
+	}
+
+	str, err := c.formatHostVolume(vol)
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Error formatting volume: %s", err))
+		return 1
+	}
+	c.Ui.Output(str)
+	return 0
+}
+
+func (c *VolumeStatusCommand) listHostVolumes(client *api.Client, nodeID, nodePool string) int {
+	vols, _, err := client.HostVolumes().List(&api.HostVolumeListRequest{
+		NodeID:   nodeID,
+		NodePool: nodePool,
+	}, nil)
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Error querying volumes: %s", err))
+		return 1
+	}
+
+	str, err := c.formatHostVolumes(vols)
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Error formatting volumes: %s", err))
+		return 1
+	}
+	c.Ui.Output(str)
+
+	return 0
+}
+
+func (c *VolumeStatusCommand) getByPrefix(client *api.Client, prefix string) (*api.HostVolumeStub, []*api.HostVolumeStub, error) {
+	vols, _, err := client.HostVolumes().List(nil, &api.QueryOptions{
+		Prefix:    prefix,
+		Namespace: c.namespace,
+	})
+
+	if err != nil {
+		return nil, nil, fmt.Errorf("error querying volumes: %s", err)
+	}
+	switch len(vols) {
+	case 0:
+		return nil, nil, fmt.Errorf("no volumes with prefix or ID %q found", prefix)
+	case 1:
+		return vols[0], nil, nil
+	default:
+		// search for exact matches to account for multiple exact ID or name
+		// matches across namespaces
+		var match *api.HostVolumeStub
+		exactMatchesCount := 0
+		for _, vol := range vols {
+			if vol.ID == prefix || vol.Name == prefix {
+				exactMatchesCount++
+				match = vol
+			}
+		}
+		if exactMatchesCount == 1 {
+			return match, nil, nil
+		}
+		return nil, vols, nil
```

Comment:
If you do that, the caller has to length-check the return value for (0, 1, many), which just repeats the logic we've done here and doesn't improve the ergonomics much. 

Think of it not as `nil, nil` but as `(obj, nil), nil` or `(nil, obj), nil`. This return value is the same as what we do in all the other places in the CLI where we do prefix-matching to find a single object. (There's even a generic helper for this `getByPrefix[T]`, but we can't use it here because the query function doesn't quite match the type param there.)